### PR TITLE
[Sol->Yul] Implementing various copying of structs

### PR DESCRIPTION
**Pull request by mijovic**
Part of #8085

Separate commits for storage to memory and vise versa.

### CODE REVIEW

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: I think it is better to do it different.
I should support structs in `readFromStorage` so this would be much simpler here. Plus it would be easier to support nested structs...

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:1542]: _fromType needs to be in the name.

File: libsolidity/codegen/LValue.cpp
Comment (chriseth) [inline libsolidity/codegen/LValue.cpp:360]: This needs a changelog entry.

File: test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol]: Can you add value parameters before and after `c` so that the data pointer gets more complicated?

**Comment by chriseth** [comment]
Please add more tests, at least one for a calldata struct that is dynamically sized / encoded.

File: libsolidity/codegen/LValue.cpp
Comment (chriseth) [inline libsolidity/codegen/LValue.cpp:364]: Please add assertion about the size on the stack of the source type.

File: test/libsolidity/semanticTests/structs/struct_storage_to_memory.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/struct_storage_to_memory.sol:15]: This way, the storage slot of `x` is zero and because of that, the test can accidentally succeed even though we did not really use the storage slot. Please add a "weirdly" sized array above `x` like `uint[37]` to make its storage slot non-trivial.

Comment (chriseth) [reply test/libsolidity/semanticTests/structs/struct_storage_to_memory.sol:15]: The same argument actually holds for calldata inputs.

File: libsolidity/codegen/LValue.cpp
Comment (chriseth) [inline libsolidity/codegen/LValue.cpp:371]: This is not relevant anymore, all of them have to.

Comment (chriseth) [reply libsolidity/codegen/LValue.cpp:371]: So maybe we don't need the loop anymore at all?

File: libsolidity/codegen/YulUtilFunctions.h
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.h]: The documentation here says this function is only meant to be used for value types.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.h]: Would it be more complicated to use special functions for non-value types? I think the separation makes sense because non-value types are much more complicated and need to allocate.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: If we process the members individually anyway, then we maybe can also use calldata member access instead of full abi decoding to memory.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Can you assign the value to a temporary local variable? You can introduce a block for each member, then you can just use a fixed name for it.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
					("hasOffset", !isStruct)
```

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: What about arrays?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I agree with that. If I remember well, I think calldata member access is not implemented yet (at least that's what #8085 says). But I can implement calldata member access, shouldn't be too complicated?

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: Totally makes sense, will introduce another function for non-value types

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Sure, then it's probably better to start with calldata member access.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Actually it is already implemented by you (#9177)

File: libsolidity/codegen/LValue.cpp
Comment (mijovic) [reply libsolidity/codegen/LValue.cpp:364]: Added both for source and target type

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I think there is no support for arrays here. I planned to add that in separate PR, to not over-complicate this one.

File: libsolidity/codegen/LValue.cpp
Comment (mijovic) [reply libsolidity/codegen/LValue.cpp:371]: I did call yul function only for copying from calldata to storage. I think we should leave the rest of function as is, and maybe later when we finalize all of the conversions in sol2Yul call for others. So, that is why I prefer not to change old code here. Also, diff looks really weird here, as I only added part related to calldata.

I think we should leave the loop as it is related to something else, not the purpose of the calldata to storage part...

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: I added two functions which are called via old function depending on if type is value or reference.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Indeed, it was really ugly line

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `fcb9f4ed2cf7e920009d9bb7a0141e1a4a5aca45`:
```
Error: Trailing whitespace found:
 Changelog.md:4: CMakeLists.txt CODE_OF_CONDUCT.md CODING_STYLE.md CONTRIBUTING.md Changelog.md LICENSE.txt README.md ReleaseChecklist.md SECURITY.md appveyor.yml cmake codecov.yml docs libevmasm liblangutil libsmtutil libsolc libsolidity libsolutil libyul scripts snap solc test tools Code generator: Implementing copying structs from calldata to storage. 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `73efa027200923bb79ee8b83d0dccdd1c9a1ae82`:
```
Coding style error:
 libsolidity/codegen/YulUtilFunctions.cpp:1274: if(_type.isValueType())

```
Please check that your changes are working as intended.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (stackenbotten) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Coding style error

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Fixed as suggested

File: test/libsolidity/semanticTests/structs/struct_memory_to_storage.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/struct_memory_to_storage.sol:16]: Please also add a buffer here.

File: test/libsolidity/semanticTests/structs/struct_copy_via_local.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/struct_copy_via_local.sol:16]: Can you also add a storage array for this test - just to be sure?

File: test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol:9]: Please add storage array.

File: test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/structs/calldata/calldata_struct_to_storage.sol]: Can you make this uint64?

File: libsolidity/codegen/LValue.cpp
Comment (chriseth) [reply libsolidity/codegen/LValue.cpp:371]: ok!

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: This should all move inside the lazy generator.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Shouldn't the name mention "reference type"?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Why didn't you use "conversion"?

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: The main difference about value types is that reading clearly always goes to the stack. For reference types, it can either go to memory or storage. It is clear that "from storage" implies "to memory" but this is less clear for "from calldata".

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Actually it turns out that the conversion function from calldata to memory is already implemented for structs.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: It is the same struct, so it should have the same members.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Arrays also do not have offsets.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: It might be that we also have to follow the tail pointer.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Oh, and validate that it is in range!

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Ah, just saw that you also modified the conversion function. I will take another look tomorrow!

**Comment by mijovic** [comment]
@chriseth I fixed some of the comments. 
Probably it was bad idea to put all of the conversions in one PR. Once you check again, if it is too hard, I can split it to one PR per converion.

**Comment by chriseth** [comment]
Tests are failing.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I think this function should not have `_offset` parameter at all (reference types have implicit offset 0), removed it now.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
	solUnimplementedAssert(_type.category() == Type::Category::Struct, "");
```
This will handle arrays as well sooner or later, won't it?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Is the criterion here really ``isStruct`` or should this rather be ``isReferenceType``?
Going forward struct members that are arrays will end up here, won't they?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Hm - this will break for function pointers... on the other hand they are unimplemented and break at a lot of places anyways I think...
Still the "proper" way of doing this would probably be to replace ``memberValue`` by ``<memberValues>`` and make that a comma separated suffixed list of variables based on the stack size of the member type...
This will also be needed, once dynamic calldata arrays as array members are supported...

File: libsolidity/codegen/YulUtilFunctions.h
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	/// @returns a function that reads a reference type from storage to memory (performing a deep copy).
```
If I understood this correctly :-)?

File: libsolidity/codegen/YulUtilFunctions.h
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.h]: I'm not quite happy with this signature... if we want to keep it like this, we should document ``_fromType``.
IIUC ``_fromType`` currently ``_fromType`` should be ``nullptr`` for all value types, but has to be a valid reference type, whenever ``_toType`` is a reference type...

Maybe the safest option would be to change the signature for this to ``updateStorageValueFunction(Type const& _fromType, Type const& _toType, ...)`` and assert ``_fromType == _toType`` if ``_toType.isValueType()``? And if ``_toType`` is a reference type maybe even also assert that ``toReferenceType.copyForLocationIfReference(&_fromType)`` is not null and the same as ``_toType``?

On the call sites of ``updateStorageValueFunction`` I guess we'll either have a type available to use as ``_fromType``, or we can just pass the same type as ``_toType`` to it...

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:1546]: See also my comment about this function in the header, so maybe we'll want to change this generally... but here we should probably assert ``solAssert(!_fromType || *_fromType == _toType, "");`` or something like that?

**Review by ekpyron — CHANGES_REQUESTED** [review]
Some comments. Did you check what happens for structs that have function pointers as members in various cases? It's fine to leave that largely unimplemented for now, but it should at least properly assert... or we just fix it now :-)...

File: libsolidity/codegen/YulUtilFunctions.h
Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.h]: (or put differently: what "target type" are you talking about :-)?)

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Updated

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: You are right, will update like suggested. :-)

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: I updated signature to accept `Type const& _fromType, Type const& _toType, ...`. Also added asserts for both value and reference types.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Fixed

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:1547]: So this will actually be called for distinct, but convertible types? Then maybe we should spend a second thinking about whether there should be some conversion functions involved in the process here :-)...

Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.cpp:1547]: Do you have an example case for which this is called with ``_fromType.isImplicitlyConvertibleTo(_toType)``, but with ``_fromType != _toType``?

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:1547]: There was few tests with integer and rational literals failed test for types being equal. So, this can be called for implicitly convertible value types...

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:1547]: For example: `uint x = 42;`

Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.cpp:1547]: Ok, that makes sense. Ok, I can't come up with a case that would be problematic here, and I guess the ``<prepare>`` and ``<update>`` below actually do some kind of conversion anyways, so this should be fine.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: This is quite a long line :-)... and again once we get to function types, this will need to be split anyways - I think in the end we'll need something like:
```
{
  let <memberValues> := <readFromStorage>(add(slot, <memberSlotDiff>)<?hasOffset>, <memberStorageOffset></hasOffset>)
  <writeToMemory>(add(value, <memberMemoryOffset>), <memberValues>)
}
```
(Using ``YulUtilFunctions::writeToMemoryFunction`` for ``writeToMemory`` instead of a plain ``mstore``)
Similarly to the ``update`` case... but we could also postpone that for implementing function types instead...

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Let's update it in this PR :-)

**Review by ekpyron — DISMISSED** [review]
So pending the remaining split into potentially multiple variables that will be required for function types, but which we could also postpone until then, I think this is fine.
We should make sure to have good test coverage with nested arrays and structs, etc, with the upcoming next step of also supporting arrays for all of this.

**Comment by mijovic** [comment]
> So pending the remaining split into potentially multiple variables that will be required for function types, but which we could also postpone until then, I think this is fine.
> We should make sure to have good test coverage with nested arrays and structs, etc, with the upcoming next step of also supporting arrays for all of this.

I opened issues regarding function types and arrays, linking them here for reference: #9725 #9727

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
	solAssert(structType.location() == DataLocation::Memory, "");
```
Calldata is read-only, so we can't read into it - it is already in storage, so reading to storage would be a no-op (for creating a copy in storage, we'd need to know which slot to copy to) - so this *has* to be memory, hasn't it?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Maybe it's safer to change this to
```suggestion
		"read_from_storage_value_type_dynamic" +
```
now?

**Review by ekpyron — DISMISSED** [review]
Two more minor optional comments - but would be fine as is as well!

**Review by ekpyron — APPROVED** [review]
(Still :-)) looks good to me! But since https://github.com/ethereum/solidity/pull/9717 is already merged, we may have to wait until after the release with merging this.

**Comment by mijovic** [comment]
> (Still :-)) looks good to me! But since #9717 is already merged, we may have to wait until after the release with merging this.

I agree. I would wait as well :-)

**Pull request merged**
